client/allocator/src/lib.rs

```diff
@@ -27,3 +27,25 @@ mod freeing_bump;
 
 pub use error::Error;
 pub use freeing_bump::{AllocationStats, FreeingBumpHeapAllocator};
+
+/// Grants access to the memory for the allocator.
+///
+/// Memory of wasm is allocated in pages. A page has a constant size of 64KiB. The maximum allowed
+/// memory size as defined in the wasm specification is 4GiB (65536 pages).
+pub trait Memory {
+	/// Run the given closure `run` and grant it write access to the raw memory.
+	fn with_access_mut<R>(&mut self, run: impl FnOnce(&mut [u8]) -> R) -> R;
+	/// Run the given closure `run` and grant it read access to the raw memory.
+	fn with_access<R>(&self, run: impl FnOnce(&[u8]) -> R) -> R;
+	/// Grow the memory by `additional` pages.
+	fn grow(&mut self, additional: u32) -> Result<(), ()>;
```

Comment:
If you use `Result` it's immediately obvious that the function can error out, and that `Result::Err` is an error. And you'll get a warning if the `Result` is not handled (since it's marked with `#[must_use]`). With `bool` there's no such convention.

And of course with `Result` you can use e.g. `map_err` and `?`.

Reply:
Good question, kind of  taste? I mean it is a failure and a simple `bool` feels too much like C, but semantically it doesn't make any difference.